## Dataset loading

In [112]:
# Loading the dataset using pandas
import pandas as pd
import numpy as np
df = pd.read_csv("../DATA/RAW/telecom_raw.csv")
df.shape

(7043, 21)

In [113]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [114]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [115]:
# converting from object dataype to numeric for better analysis
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')


In [116]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


,SeniorCitizen,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000,7032.000000
mean,0.162147,32.371149,64.761692,2283.300441
std,0.368612,24.559481,30.090047,2266.771362
min,0.000000,0.000000,18.250000,18.800000
25%,0.000000,9.000000,35.500000,401.450000
50%,0.000000,29.000000,70.350000,1397.475000
75%,0.000000,55.000000,89.850000,3794.737500
max,1.000000,72.000000,118.750000,8684.800000


In [117]:
# checking if there are any missing values 
df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [118]:
# Imputing missing values in TotalCharges column with the median charges 
df['TotalCharges'] = df['TotalCharges'].fillna(df['TotalCharges'].median())


In [119]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [120]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [121]:
# Renaming columns according to snake casing for better readability and documentation
df.columns = (
    df.columns
      .str.replace('([a-z0-9])([A-Z])', r'\1_\2', regex=True)  
      .str.replace(' ', '_')                                  
      .str.lower()
)


In [122]:
df.columns

Index(['customer_id', 'gender', 'senior_citizen', 'partner', 'dependents',
       'tenure', 'phone_service', 'multiple_lines', 'internet_service',
       'online_security', 'online_backup', 'device_protection', 'tech_support',
       'streaming_tv', 'streaming_movies', 'contract', 'paperless_billing',
       'payment_method', 'monthly_charges', 'total_charges', 'churn'],
      dtype='object')

In [123]:
# creating a cleaned telecome datase for further feature engineering
df.to_csv("telecom_cleaned.csv", index=False)


## Feature Engineering

In [125]:
df['churn_flag'] = df['churn'].map({'Yes': 1, 'No': 0})


In [126]:
df['customer_lifetime_value'] = df['monthly_charges'] * df['tenure']


In [127]:
df['revenue_at_risk'] = np.where(
    df['churn_flag'] == 1,
    df['monthly_charges'],
    0
)


In [128]:
df['tenure_bucket'] = pd.cut(
    df['tenure'],
    bins=[0, 6, 12, 24, np.inf],
    labels=['0-6 months', '6-12 months', '12-24 months', '24+ months'],
    include_lowest=True
)


In [129]:
contract_risk_map = {
    'Month-to-month': 'High',
    'One year': 'Medium',
    'Two year': 'Low'
}

df['contract_risk'] = df['contract'].map(contract_risk_map)


In [130]:
high_charge_threshold = df['monthly_charges'].quantile(0.75)

df['price_sensitivity_flag'] = np.where(
    (df['monthly_charges'] >= high_charge_threshold) &
    (df['tenure'] < 12),
    1,
    0
)


In [131]:
feature_cols = [
    'customer_lifetime_value',
    'revenue_at_risk',
    'tenure_bucket',
    'contract_risk',
    'price_sensitivity_flag'
]

df[feature_cols].head()


,customer_lifetime_value,revenue_at_risk,tenure_bucket,contract_risk,price_sensitivity_flag
0,29.85,0.00,0-6 months,High,0
1,1936.30,0.00,24+ months,Medium,0
2,107.70,53.85,0-6 months,High,0
3,1903.50,0.00,24+ months,Medium,0
4,141.40,70.70,0-6 months,High,0


In [132]:
df[feature_cols].info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   customer_lifetime_value  7043 non-null   float64 
 1   revenue_at_risk          7043 non-null   float64 
 2   tenure_bucket            7043 non-null   category
 3   contract_risk            7043 non-null   object  
 4   price_sensitivity_flag   7043 non-null   int32   
dtypes: category(1), float64(2), int32(1), object(1)
memory usage: 199.8+ KB


In [154]:
df.to_csv("telecom_feature_engineered_cleaned.csv", index=False)


In [159]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 27 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   customer_id              7043 non-null   object  
 1   gender                   7043 non-null   object  
 2   senior_citizen           7043 non-null   int64   
 3   partner                  7043 non-null   object  
 4   dependents               7043 non-null   object  
 5   tenure                   7043 non-null   int64   
 6   phone_service            7043 non-null   object  
 7   multiple_lines           7043 non-null   object  
 8   internet_service         7043 non-null   object  
 9   online_security          7043 non-null   object  
 10  online_backup            7043 non-null   object  
 11  device_protection        7043 non-null   object  
 12  tech_support             7043 non-null   object  
 13  streaming_tv             7043 non-null   object  
 14  streamin

## Exploratory Data Analysis (EDA)

### Spending Behavior


In [162]:
df[['tenure', 'monthly_charges', 'customer_lifetime_value']].describe()


,tenure,monthly_charges,customer_lifetime_value
count,7043.000000,7043.000000,7043.000000
mean,32.371149,64.761692,2279.581350
std,24.559481,30.090047,2264.729447
min,0.000000,18.250000,0.000000
25%,9.000000,35.500000,394.000000
50%,29.000000,70.350000,1393.600000
75%,55.000000,89.850000,3786.100000
max,72.000000,118.750000,8550.000000


**Insight:**  
Customer tenure shows a wide spread (0 to 72 months), indicating the presence of both new and long-term customers, making lifecycle and cohort-based churn analysis meaningful.

Monthly charges vary significantly across customers, enabling effective analysis of price sensitivity and its impact on churn behavior.

Customer Lifetime Value (CLV) is highly skewed, with a small segment of customers contributing disproportionately to overall revenue, highlighting the importance of value-based retention strategies.


### Overall Churn and Revenue Impact Summary

In [204]:
overall_summary['value'] = overall_summary.apply(
    lambda row: round(row['value'] * 100, 2) if 'Rate' in row['metric'] else round(row['value'], 2),
    axis=1
)

overall_summary


,metric,value
0,Overall Churn Rate,26.54
1,Total Revenue at Risk,139130.85


**Insight:**  
An overall churn rate of approximately 26.5% corresponds to a monthly revenue loss of about 139,131, underscoring the significant financial impact of customer attrition.

### Tenure Cohort Summary: Customer Distribution, Churn, and Revenue Impact

In [193]:
tenure_cohort_summary = (
    df.groupby('tenure_bucket', observed=True)
      .agg(
          customer_share=('customer_id', 'count'),
          revenue_at_risk=('revenue_at_risk', 'sum'),
          churn_rate=('churn_flag', 'mean')
      )
)

# Convert counts to percentage
tenure_cohort_summary['customer_share_pct'] = (
    tenure_cohort_summary['customer_share'] /
    tenure_cohort_summary['customer_share'].sum()
)

# Sort by revenue impact
tenure_cohort_summary = tenure_cohort_summary.sort_values(
    by='revenue_at_risk',
    ascending=False
)

tenure_cohort_summary


,customer_share,revenue_at_risk,churn_rate,customer_share_pct
tenure_bucket,,,,
0-6 months,1481,49896.10,0.529372,0.210280
24+ months,3833,47094.95,0.140360,0.544228
12-24 months,1024,23081.65,0.287109,0.145393
6-12 months,705,19058.15,0.358865,0.100099


**Insight:**  
Although customers in their first 6 months represent only ~21% of the customer base, they account for the highest churn rate (~53%) and the largest share of revenue at risk, making early-stage retention the most critical business priority.

### Contract Risk Summary: Customer Distribution, Churn, and Revenue Impact

In [197]:
contract_risk_summary = (
    df.groupby('contract_risk', observed=True)
      .agg(
          customer_count=('customer_id', 'count'),
          churn_rate=('churn_flag', 'mean'),
          revenue_at_risk=('revenue_at_risk', 'sum')
      )
)

# Convert count to percentage
contract_risk_summary['customer_share_pct'] = (
    contract_risk_summary['customer_count'] /
    contract_risk_summary['customer_count'].sum()
)

# Sort by churn severity
contract_risk_summary = contract_risk_summary.sort_values(
    by='churn_rate',
    ascending=False
)

contract_risk_summary


,customer_count,churn_rate,revenue_at_risk,customer_share_pct
contract_risk,,,,
High,3875,0.427097,120847.10,0.550192
Medium,1473,0.112695,14118.45,0.209144
Low,1695,0.028319,4165.30,0.240664


**Insight:**  
Month-to-month customers represent over half of the customer base and exhibit a significantly higher churn rate (~43%), contributing to the majority of revenue at risk and highlighting contract type as a critical retention lever.